### PR TITLE
updated to functionsV2

### LINKED ISSUE #124: Migrate onVoteRequestUpdate to functions v2
https://firebase.google.com/docs/functions/2nd-gen-upgrade

### LINKED ISSUE #125: Migrate onMessageUpdate to functions v2
https://firebase.google.com/docs/functions/2nd-gen-upgrade

### LINKED ISSUE #126: Migrate healthcheck, onInstanceUpdate, onInstanceDelete, onMessageWrite to functions v2
https://firebase.google.com/docs/functions/2nd-gen-upgrade

### PR DESCRIPTION
**Pull request by sarge1989**
# Issues
- Resolves #124 
- Resolves #125 
- Resolves #126 

# Summary
- Converted to functions v2 as per https://firebase.google.com/docs/functions/2nd-gen-upgrade
- Since the whatsapp webhook will be taken down, might as well do this now and resolve the technical debt.

### CODE REVIEW

File: functions/src/definitions/eventHandlers/onInstanceCreate.ts [hunk 1]
```diff
@@ -32,10 +33,13 @@ const onInstanceCreate = functions
...
-  })
-  .firestore.document("/messages/{messageId}/instances/{instanceId}")
-  .onCreate(async (snap, context) => {
-    // Grab the current value of what was written to Firestore.
+  },
+  async (event) => {
+    const snap = event.data
+    if (!snap) {
+      console.log("No data associated with the event")
+      return
```
Comment (btjl) [inline functions/src/definitions/eventHandlers/onInstanceCreate.ts:41]:
```suggestion
      return Promise.resolve()
```

File: functions/src/definitions/eventHandlers/onInstanceCreate.ts [hunk 2]
```diff
@@ -21,9 +22,9 @@ if (!admin.apps.length) {
   admin.initializeApp()
 }
 
-const onInstanceCreate = functions
-  .region("asia-southeast1")
-  .runWith({
+const onInstanceCreateV2 = onDocumentCreated(
+  {
+    document: "messages/{messageId}/instances/{instanceId}",
```
Comment (btjl) [inline functions/src/definitions/eventHandlers/onInstanceCreate.ts:27]:
```suggestion
    document: "/messages/{messageId}/instances/{instanceId}",
```

File: functions/src/definitions/eventHandlers/onMessageUpdate.ts [hunk 3]
```diff
@@ -2,21 +2,24 @@ import * as functions from "firebase-functions"
...
 import { Timestamp } from "firebase-admin/firestore"
 import { rationaliseMessage, anonymiseMessage } from "../common/genAI"
+import { onDocumentUpdated } from "firebase-functions/v2/firestore"
 
-const onMessageUpdate = functions
-  .region("asia-southeast1")
-  .runWith({
+const onMessageUpdateV2 = onDocumentUpdated(
+  {
+    document: "messages/{messageId}",
```
Comment (btjl) [inline functions/src/definitions/eventHandlers/onMessageUpdate.ts:9]:
```suggestion
    document: "/messages/{messageId}",
```

File: functions/src/definitions/eventHandlers/onInstanceUpdate.ts [hunk 4]
```diff
@@ -4,30 +4,35 @@ import {
...
     secrets: ["TYPESENSE_TOKEN", "ML_SERVER_TOKEN"],
-  })
-  .firestore.document("/messages/{messageId}/instances/{instanceId}")
-  .onUpdate(async (change, context) => {
+  },
+  async (event) => {
     // Grab the current value of what was written to Firestore.
-    const before = change.before.data()
-    const after = change.after.data()
+    if (!event?.data?.before || !event?.data?.before) {
```
Comment (btjl) [inline functions/src/definitions/eventHandlers/onInstanceUpdate.ts:16]: This is testing the same thing

File: functions/src/definitions/eventHandlers/onVoteRequestUpdate.ts [hunk 5]
```diff
@@ -19,29 +20,31 @@ if (!admin.apps.length) {
...
-  })
-  .firestore.document("/messages/{messageId}/voteRequests/{voteRequestId}")
-  .onUpdate(async (change, context) => {
+  },
+  async (event) => {
     // Grab the current value of what was written to Firestore.
-    const before = change.before.data()
-    const docSnap = change.after
-    const after = docSnap.data()
+    if (!event?.data?.before || !event?.data?.before) {
```
Comment (btjl) [inline functions/src/definitions/eventHandlers/onVoteRequestUpdate.ts:30]: This is testing the same thing

File: functions/src/definitions/eventHandlers/onVoteRequestUpdate.ts [hunk 6]
```diff
@@ -19,29 +20,31 @@ if (!admin.apps.length) {
...
+      return Promise.resolve()
+    }
+    const before = event.data.before.data()
```
Comment (btjl) [inline functions/src/definitions/eventHandlers/onVoteRequestUpdate.ts:33]: I think a more descriptive name for `before` and `after` would be beneficial

File: functions/src/definitions/eventHandlers/onInstanceCreate.ts (hunk 2 above)
Comment (sarge1989) [reply functions/src/definitions/eventHandlers/onInstanceCreate.ts:27]: The examples in the documentation seem to go without the leading slash so I went with it!

File: functions/src/definitions/eventHandlers/onInstanceUpdate.ts (hunk 4 above)
Comment (sarge1989) [reply functions/src/definitions/eventHandlers/onInstanceUpdate.ts:16]: fixed

File: functions/src/definitions/eventHandlers/onMessageUpdate.ts (hunk 3 above)
Comment (sarge1989) [reply functions/src/definitions/eventHandlers/onMessageUpdate.ts:9]: The examples in the documentation seem to go without the leading slash so I went with it!

File: functions/src/definitions/eventHandlers/onVoteRequestUpdate.ts [hunk 7]
```diff
@@ -19,29 +20,31 @@ if (!admin.apps.length) {
 
 const db = admin.firestore()
 
-const onVoteRequestUpdate = functions
-  .region("asia-southeast1")
-  .runWith({
+const onVoteRequestUpdateV2 = onDocumentUpdated(
+  {
+    document: "messages/{messageId}/voteRequests/{voteRequestId}",
```
[the remark this answers is not in the document]
Comment (sarge1989) [reply functions/src/definitions/eventHandlers/onVoteRequestUpdate.ts:25]: The examples in the documentation seem to go without the leading slash so I went with it!

File: functions/src/definitions/eventHandlers/updateTypesense.ts [hunk 8]
```diff
@@ -5,16 +5,19 @@ import {
...
+import { onDocumentWritten } from "firebase-functions/v2/firestore"
 
-const onMessageWrite = functions
-  .region("asia-southeast1")
-  .runWith({ secrets: ["TYPESENSE_TOKEN"] })
-  .firestore.document("/messages/{messageId}")
-  .onWrite(async (snap, context) => {
+const onMessageWriteV2 = onDocumentWritten(
+  {
+    document: "messages/{messageId}",
```
[the remark this answers is not in the document]
Comment (sarge1989) [reply functions/src/definitions/eventHandlers/updateTypesense.ts:12]: The examples in the documentation seem to go without the leading slash so I went with it!

File: functions/src/definitions/eventHandlers/updateTypesense.ts [hunk 9]
```diff
@@ -5,16 +5,19 @@ import {
...
+    secrets: ["TYPESENSE_TOKEN"],
+  },
+  async (event) => {
     // Grab the current value of what was written to Firestore.
-    const id = context.params.messageId
+    const id = event.params.messageId
     // If the message is deleted, delete from the Typesense index
-    if (!snap.after.exists) {
+    const messageSnap = event?.data?.after
+    if (messageSnap == undefined || !messageSnap.exists) {
```
[the remark this answers is not in the document]
Comment (sarge1989) [reply functions/src/definitions/eventHandlers/updateTypesense.ts:20]: changed!

File: functions/src/definitions/eventHandlers/onVoteRequestUpdate.ts (hunk 5 above)
Comment (sarge1989) [reply functions/src/definitions/eventHandlers/onVoteRequestUpdate.ts:30]: fixed

File: functions/src/definitions/eventHandlers/onInstanceCreate.ts (hunk 1 above)
Comment (sarge1989) [reply functions/src/definitions/eventHandlers/onInstanceCreate.ts:41]: Fixed

File: functions/src/definitions/eventHandlers/onVoteRequestUpdate.ts (hunk 6 above)
Comment (sarge1989) [reply functions/src/definitions/eventHandlers/onVoteRequestUpdate.ts:33]: Ok changed!

File: functions/src/definitions/eventHandlers/onMessageUpdate.ts [hunk 10]
```diff
@@ -2,21 +2,24 @@ import * as functions from "firebase-functions"
...
-  })
-  .firestore.document("/messages/{messageId}")
-  .onUpdate(async (change, context) => {
+  },
+  async (event) => {
     // Grab the current value of what was written to Firestore.
-    const before = change.before
-    const after = change.after
+    const before = event?.data?.before
+    const after = event?.data?.after
```
Comment (btjl) [inline functions/src/definitions/eventHandlers/onMessageUpdate.ts:19]: Can apply the more descriptive names here as well

File: functions/src/definitions/eventHandlers/onInstanceUpdate.ts [hunk 11]
```diff
@@ -4,30 +4,35 @@ import {
...
+  },
+  async (event) => {
     // Grab the current value of what was written to Firestore.
-    const before = change.before.data()
-    const after = change.after.data()
+    if (!event?.data?.before || !event?.data?.after) {
+      return Promise.resolve()
+    }
+    const before = event.data.before.data()
+    const snap = event.data.after
```
Comment (btjl) [inline functions/src/definitions/eventHandlers/onInstanceUpdate.ts:20]: Can apply the more descriptive names here as well

File: functions/src/definitions/eventHandlers/onInstanceUpdate.ts [hunk 12]
```diff
@@ -4,30 +4,35 @@ import {
...
 } from "../common/typesense/collectionOperations"
 import { getEmbedding } from "../common/machineLearningServer/operations"
+import { onDocumentUpdated } from "firebase-functions/v2/firestore"
 
-const onInstanceUpdate = functions
-  .region("asia-southeast1")
-  .runWith({
+const onInstanceUpdateV2 = onDocumentUpdated(
+  {
+    document: "/messages/{messageId}/instances/{instanceId}",
```
Comment (btjl) [inline functions/src/definitions/eventHandlers/onInstanceUpdate.ts:11]:
```suggestion
    document: "messages/{messageId}/instances/{instanceId}",
```

To standardise with no leading slash

File: functions/src/definitions/eventHandlers/updateTypesense.ts [hunk 13]
```diff
@@ -5,16 +5,19 @@ import {
...
+    document: "messages/{messageId}",
+    secrets: ["TYPESENSE_TOKEN"],
+  },
+  async (event) => {
     // Grab the current value of what was written to Firestore.
-    const id = context.params.messageId
+    const id = event.params.messageId
     // If the message is deleted, delete from the Typesense index
-    if (!snap.after.exists) {
+    const messageSnap = event?.data?.after
```
Comment (btjl) [inline functions/src/definitions/eventHandlers/updateTypesense.ts:19]:
```suggestion
    const postMessageSnap = event?.data?.after
```

Similar to `postChangeData`

File: functions/src/definitions/eventHandlers/onInstanceUpdate.ts (hunk 12 above)
Comment (sarge1989) [reply functions/src/definitions/eventHandlers/onInstanceUpdate.ts:11]: Done

File: functions/src/definitions/eventHandlers/onMessageUpdate.ts (hunk 10 above)
Comment (sarge1989) [reply functions/src/definitions/eventHandlers/onMessageUpdate.ts:19]: Ok done

File: functions/src/definitions/eventHandlers/onInstanceUpdate.ts (hunk 11 above)
Comment (sarge1989) [reply functions/src/definitions/eventHandlers/onInstanceUpdate.ts:20]: Ok done

File: functions/src/definitions/eventHandlers/updateTypesense.ts (hunk 13 above)
Comment (sarge1989) [reply functions/src/definitions/eventHandlers/updateTypesense.ts:19]: ok addressed this

**Comment by btjl** [comment]
LGTM!

**Review by btjl — APPROVED** [review]

**Pull request merged**